tests/slist_ut/slist_ut.c

```diff
@@ -0,0 +1,1114 @@
+MOCK_FUNCTION_END(0);
+
+MOCK_FUNCTION_WITH_CODE(, int, test_action_function_2, PSINGLYLINKEDLIST_ENTRY, list_entry, const void*, action_context, bool*, continueProcessing);
+simpleItem* item = CONTAINING_RECORD(list_entry, simpleItem, link);
+*continueProcessing = item->index != 3;
+MOCK_FUNCTION_END(0);
+
+MOCK_FUNCTION_WITH_CODE(, int, test_action_function_fail, PSINGLYLINKEDLIST_ENTRY, list_entry, const void*, action_context, bool*, continueProcessing);
+*continueProcessing = true;
+simpleItem* item = CONTAINING_RECORD(list_entry, simpleItem, link);
+(int)item->index;
+if (item->index == 3)
+{
+    return MU_FAILURE;
+}
+return 0;
+MOCK_FUNCTION_END();
+
+#undef ENABLE_MOCKS
+
+static TEST_MUTEX_HANDLE test_serialize_mutex;
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+{
+    ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
+}
+
+BEGIN_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
+
+TEST_SUITE_INITIALIZE(TestClassInitialize)
+{
+    int result;
+
+    test_serialize_mutex = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(test_serialize_mutex);
+
+    umock_c_init(on_umock_c_error);
+
+    result = umocktypes_bool_register_types();
+    ASSERT_ARE_EQUAL(int, 0, result);
+
+    result = umocktypes_charptr_register_types();
+    ASSERT_ARE_EQUAL(int, 0, result);
+
+    REGISTER_UMOCK_ALIAS_TYPE(PSINGLYLINKEDLIST_ENTRY, void*);
+}
+
+TEST_SUITE_CLEANUP(TestClassCleanup)
+{
+    umock_c_deinit();
+    TEST_MUTEX_DESTROY(test_serialize_mutex);
+}
+
+TEST_FUNCTION_INITIALIZE(TestMethodInitialize)
+{
+    if (TEST_MUTEX_ACQUIRE(test_serialize_mutex))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+}
+
+TEST_FUNCTION_CLEANUP(TestMethodCleanup)
+{
+    umock_c_reset_all_calls();
+    TEST_MUTEX_RELEASE(test_serialize_mutex);
+}
+
+/*slist_initialize*/
+
+/* Tests_SRS_SLIST_07_001: [ If `list_head` is `NULL`, `slist_initialize` shall fail and return `false`. ]*/
+TEST_FUNCTION(slist_initialize_with_NULL_fails)
+{
+    // arrange
+
+    // act
+    bool result = slist_initialize(NULL);
+
+    // assert
+    ASSERT_ARE_EQUAL(bool, false, result);
+}
+
+/* Tests_SRS_SLIST_07_002: [ `slist_initialize` shall initialize the `next` pointer in `list_head` points to `NULL`and return `true` on success. ]*/
+TEST_FUNCTION(slist_initialize_succeeds)
+{
+    // arrange
+    SINGLYLINKEDLIST_ENTRY head;
+
+    // act
+    bool result = slist_initialize(&head);
+
+    // assert
+    ASSERT_ARE_EQUAL(void_ptr, NULL, head.next);
+    ASSERT_ARE_EQUAL(bool, true, result);
+}
+
+/*slist_is_empty*/
+
+/* Tests_SRS_SLIST_07_038: [ If `list_head` is `NULL`, `slist_is_empty` shall fail and return `true`. ]*/
+TEST_FUNCTION(slist_is_empty_with_NULL_list_fails)
+{
+    // arrange
+
+    // act
+    bool result = slist_is_empty(NULL);
+
+    // assert
+    ASSERT_ARE_EQUAL(bool, true, result);
```

Comment:
Technically this is not an empty list case.
It is an invalid args case.
So I suggest returning an enum that has 3 values: EMPTY, NOT_EMPTY, INVALID_ARGS

Reply:
done